velox/functions/prestosql/window/Ntile.cpp

```diff
@@ -0,0 +1,242 @@
+    partition_ = partition;
+    partitionOffset_ = 0;
+    numPartitionRows_ = partition->numRows();
+
+    if (numFixedBuckets_.has_value()) {
+      auto numBuckets = numFixedBuckets_.value();
+      // If there are more buckets than partition rows, then the output bucket
+      // number is the same as the row number so no further computation is
+      // needed.
+      fixedBucketsMoreThanPartition_ = numBuckets > numPartitionRows_;
+      if (!fixedBucketsMoreThanPartition_) {
+        fixedBucketMetrics_ = computeBucketMetrics(numBuckets);
+      }
+    }
+  }
+
+  void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& /*frameStarts*/,
+      const BufferPtr& /*frameEnds*/,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    int numRows = peerGroupStarts->size() / sizeof(vector_size_t);
+
+    if (bucketColumn_.has_value()) {
+      computeNtileFromColumn(numRows, resultOffset, result);
+    } else {
+      computeNtileWithConstants(numRows, resultOffset, result);
+    }
+
+    partitionOffset_ += numRows;
+  }
+
+ private:
+  // These are some intermediate values required for bucket computation when the
+  // number of rows in the partition exceeds the number of buckets.
+  struct BucketMetrics {
+    // To compute the bucket number for a row, we find the number of rows in
+    // a bucket as the (number of rows in partition) / (number of buckets).
+    int64_t rowsPerBucket;
+    // There could be some buckets with rowsPerBucket + 1 number of rows,
+    // as the partition rows might not be exactly divisible
+    // by the number of buckets. There are
+    // (number of rows in partition) % (number of buckets) such buckets.
+    int64_t bucketsWithExtraRow;
+    // When assigning bucket numbers, the first 'bucketsWithExtraRow' buckets
+    // will have (rowsPerBucket + 1) rows. This row number at this boundary is
+    // extraBucketsBoundary = bucketsWithExtraRow * (rowsPerBucket + 1). Beyond
+    // this row number in the partition, the buckets will have only
+    // rowsPerBucket number of rows. This boundary is useful when computing the
+    // bucket value.
+    int64_t extraBucketsBoundary;
+
+    int64_t computeBucketValue(vector_size_t rowNumber) const {
+      if (rowNumber < extraBucketsBoundary) {
+        return rowNumber / (rowsPerBucket + 1) + 1;
+      }
+      return (rowNumber - bucketsWithExtraRow) / rowsPerBucket + 1;
+    }
+
+    // This computes the bucket value for a fixed bucket number for a vector
+    // of rows. The vector starts at the partitionOffset index in the
+    // partition rows.
+    void computeBucketValue(
+        vector_size_t numRows,
+        vector_size_t partitionOffset,
+        vector_size_t resultOffset,
+        int64_t* rawResultValues) {
+      vector_size_t j = 0;
+      for (auto i = partitionOffset; i < extraBucketsBoundary; i++, j++) {
```

Comment:
looks like i and j are the same; consider removing one of them (use 'i' name for the remaining one).

Reply:
I have changed the code so that i goes from 0 -> numRows, whereas the index j goes from partitionOffset_ to extraBucketsBoundary (the value at which the computation changes). I can try to eliminate j, but the new code subtracting different numbers is a big ugly.